### PR TITLE
[DROOLS-6497] when updating a parent rule during the incremental comp…

### PR DESCRIPTION
**Pull request by mariofusco**
…ilation also recompile all its children

**JIRA**: https://issues.redhat.com/browse/DROOLS-6497

### CODE REVIEW

**Comment by mariofusco** [comment]
Jankins failure of `org.drools.compiler.integrationtests.TimerAndCalendarTest.testHaltAfterSomeTimeThenRestartButNoLongerHolding` is unrelated

File: drools-compiler/src/main/java/org/drools/compiler/builder/impl/KnowledgeBuilderImpl.java [hunk 1]
```diff
@@ -1258,12 +1258,38 @@ private void preProcessRules(PackageDescr packageDescr, PackageRegistry pkgRegis
...
     }
 
+    private Collection<RuleImpl> findChildrenRulesToBeRemoved(PackageDescr packageDescr, Collection<RuleImpl> rulesToBeRemoved) {
+        Collection<String> childrenRuleNamesToBeRemoved = new HashSet<>();
+        Collection<RuleImpl> childrenRulesToBeRemoved = new HashSet<>();
+        for (RuleImpl rule : rulesToBeRemoved) {
+            if (rule.hasChildren()) {
+                for (RuleImpl child : rule.getChildren()) {
+                    if (!rulesToBeRemoved.contains(child)) {
+                        // if a rule has child not already contained among the ones to be removed ...
```
Comment (lucamolteni) [inline drools-compiler/src/main/java/org/drools/compiler/builder/impl/KnowledgeBuilderImpl.java]: If a rule has a child rule not marked to be removed

File: drools-compiler/src/main/java/org/drools/compiler/builder/impl/KnowledgeBuilderImpl.java [hunk 2]
```diff
@@ -1258,12 +1258,38 @@ private void preProcessRules(PackageDescr packageDescr, PackageRegistry pkgRegis
...
+                        childrenRulesToBeRemoved.add(child);
+                        childrenRuleNamesToBeRemoved.add(child.getName());
+                        // ... remove the child rule but also add it back to the PackageDescr in order to readd it when also the parent rule will be ...
```
Comment (lucamolteni) [inline drools-compiler/src/main/java/org/drools/compiler/builder/impl/KnowledgeBuilderImpl.java]: Will be what? :D

**Review by lucamolteni — CHANGES_REQUESTED** [review]

File: drools-compiler/src/main/java/org/drools/compiler/builder/impl/KnowledgeBuilderImpl.java (hunk 2 above)
Comment (mariofusco) [reply drools-compiler/src/main/java/org/drools/compiler/builder/impl/KnowledgeBuilderImpl.java]: I meant "readd it when also the parent rule will be [readded]". Is the sentence unclear?

**Review by hellowdan — APPROVED** [review]

**Review by pibizza — APPROVED** [review]
Sounds good for me.

**Review by lucamolteni — APPROVED** [review]

**Pull request merged**